### PR TITLE
Disallow overriding initializers if IR version < 4 

### PR DESCRIPTION
**Pull request by skottmckay**
**Description**: 

* Disallow overriding an initializer via a graph input if the IR version is < 4. This enforces an implicit assumption that initializers should be treated as constant, and allows constant folding to be done on a model with an older IR version.
* Separate constant and overridable initializers so that it's clear which ones constant folding can utilize.
* Update Graph to not add all initializers to the graph inputs when the graph is manually created (i.e. not loaded from a GraphProto) and the IR version is >= 4.

**Motivation and Context**
In order to do constant folding we need to know which initializers can be treated as constant and which are overridable. All initializers were required to have a matching graph input prior to IR version 4, technically making all of them overridable. The intention however was for them to be treated as constants, and this change enforces that intent. 

The benefit of doing so is that constant folding will work for models with IR version < 4. The cost is that if someone is actually overriding an initializer they will need to update the IR version of their model to version 4 in order to keep doing so. The belief is that this is a very small subset of usage (e.g. models involving feeding in a truncated sequence) and the cost to update that small subset is warranted by the benefit of constant folding being able to be enabled on all older models without them needing an IR version update.

### CODE REVIEW

File: include/onnxruntime/core/framework/kernel_registry.h [hunk 1]
```diff
@@ -25,7 +25,7 @@ class KernelRegistry {
   // itself.
   // TODO(Task:132) Make usage of unique_ptr/shared_ptr as out param consistent
   Status TryCreateKernel(const onnxruntime::Node& node, const IExecutionProvider& execution_provider,
-                         const std::unordered_map<int, OrtValue>& initialized_tensors,
+                         const std::unordered_map<int, OrtValue>& constant_initialized_tensors,
```
Comment (KeDengMS) [inline include/onnxruntime/core/framework/kernel_registry.h]: >constant_initialized_tensors [](start = 66, length = 28)

Are these changes necessary? IMO we should minimize changes to framework if possible.

File: onnxruntime/core/framework/session_state.cc [hunk 2]
```diff
@@ -36,17 +36,32 @@ void SessionState::SetExecutionPlan(std::unique_ptr<SequentialExecutionPlan> p_s
...
   auto p = initialized_tensors_.insert({ort_value_index, ort_value});
   if (!p.second)
     return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "duplicated ort_value index:", ort_value_index,
                            ". Do you have duplicated calls to SessionState::AddInitializedTensor function?");
-  if (d != nullptr && d->f != nullptr) deleter_for_initialized_tensors_[ort_value_index] = *d;
+
+  if (d != nullptr && d->f != nullptr) {
+    deleter_for_initialized_tensors_[ort_value_index] = *d;
+  }
+
```
Comment (KeDengMS) [inline onnxruntime/core/framework/session_state.cc]: nit: trim space at line end

File: onnxruntime/core/framework/session_state_initializer.cc [hunk 3]
```diff
@@ -243,28 +245,35 @@ common::Status SaveInitializedTensors(const Env& env, const std::basic_string<PA
 
   //1. first plan the memory
   const onnxruntime::InitializedTensorSet& initialized_tensor_set = graph.GetAllInitializedTensors();
-  std::unordered_map<int, const ONNX_NAMESPACE::TensorProto*> id_to_initialized_tensor;
+  // map of ort value index to <initializer name, TensorProto>
+  using NameAndProto = std::pair<const std::string, const ONNX_NAMESPACE::TensorProto*>;
```
Comment (KeDengMS) [inline onnxruntime/core/framework/session_state_initializer.cc]: >NameAndProto [](start = 8, length = 12)

NameAndTensorProto?

Comment (KeDengMS) [inline onnxruntime/core/framework/session_state_initializer.cc]: >TensorProto [](start = 74, length = 11)

TensorProto has names for initializers, why do we need another name? Could they be different?

File: onnxruntime/core/graph/graph_utils.cc [hunk 4]
```diff
@@ -399,9 +399,24 @@ bool IsGraphInput(const Graph& graph, const NodeArg* input) {
   return std::find(graph_inputs.begin(), graph_inputs.end(), input) != graph_inputs.end();
 }
 
+bool IsConstantInitializer(const Graph& graph, const std::string& initializer_name) {
+  bool constant = true;
+
+  if (graph.CanOverrideInitializer()) {
```
Comment (KeDengMS) [inline onnxruntime/core/graph/graph_utils.cc]: >graph.CanOverrideInitializer() [](start = 6, length = 30)

For initializers in subgraph, do we allow them to be overridden?

File: onnxruntime/core/graph/graph_utils.cc [hunk 5]
```diff
@@ -399,9 +399,24 @@ bool IsGraphInput(const Graph& graph, const NodeArg* input) {
...
+    const auto& graph_inputs = graph.GetInputsIncludingInitializers();
```
Comment (KeDengMS) [inline onnxruntime/core/graph/graph_utils.cc]: >GetInputsIncludingInitializers [](start = 37, length = 30)

Might be cleaner if the function name is GetInputsIncludingOverridableInitializers(). Or maybe GetInputs would do if all overridable initializers are already in graph inputs?

File: onnxruntime/core/optimizer/constant_folding.cc [hunk 6]
```diff
@@ -37,6 +37,8 @@ Status ConstantFolding::ApplyImpl(Graph& graph, bool& modified, int graph_level)
     }
 
     // Create execution frame for executing constant nodes.
+    // NOTE: As we call AllNodeInputsAreConstant we can use the full list of initializers from
+    // graph.GetAllInitializedTensors() without filtering out overridable (i.e. non-constant) initializers
```
Comment (KeDengMS) [inline onnxruntime/core/optimizer/constant_folding.cc]: >graph.GetAllInitializedTensors() without filtering out overridable (i.e. non-constant) initializers [](start = 7, length = 99)

Seems graph.GetAllInitializedTensors() has a change in semantics, which may affect execution providers that have a different assumption. I think it's better to rename the function when semantics changing, so we'll inspect the places this is used in the PR.

File: onnxruntime/core/session/inference_session.cc [hunk 7]
```diff
@@ -561,13 +561,29 @@ common::Status InferenceSession::ValidateInputs(const std::vector<std::string>&
...
-    auto iter = input_def_map_.find(feed_names[i]);
+    const auto& feed_name = feed_names[i];
+
+    if (seen_names.insert(feed_name).second == false) {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "Duplicate name in feeds: ", feed_name);
+    }
+
+    auto iter = input_def_map_.find(feed_name);
     if (input_def_map_.end() == iter) {
+      // if IR < 4 we disallow using a graph input to override an initializer. check for this and output a nicer
```
Comment (KeDengMS) [inline onnxruntime/core/session/inference_session.cc]: >we disallow using a graph input to override an initializer [](start = 19, length = 58)

maybe add "even if the initializer name is listed explicitly in graph input"

File: include/onnxruntime/core/graph/graph.h [hunk 8]
```diff
@@ -962,7 +965,7 @@ class Graph {
   std::unordered_map<std::string, const ONNX_NAMESPACE::FunctionProto*> model_functions_;
 
   // Model IR version.
-  Version ir_version_{};
+  Version ir_version_{ONNX_NAMESPACE::Version::IR_VERSION};
```
Comment (snnn) [inline include/onnxruntime/core/graph/graph.h]: So, this is a constant expr right?

File: onnxruntime/core/session/inference_session.cc [hunk 9]
```diff
@@ -580,6 +596,32 @@ common::Status InferenceSession::ValidateInputs(const std::vector<std::string>&
       auto input_type = input_ml_value.Type();
       ORT_RETURN_IF_ERROR(CheckTypes(input_type, expected_type));
     }
+
+    // IR < 4 all entries in input_def_map_ are required.
+    // IR 4 or above we need to do a lookup to see if this was a required input
+    if (model_ir_version < 4 || required_inputs_.find(feed_name) != required_inputs_.cend()) {
```
Comment (KeDengMS) [inline onnxruntime/core/session/inference_session.cc]: Would be better if we have a util function, instead of hard- code version 4 in comments...

**Comment by KeDengMS** [comment]
> CodeFlow for GitHub does not download executable files. If you trust this file and would like to run it, please download it yourself through GitHub.

should the model file named .pb like others in the folder?

---
Refers to: onnxruntime/test/testdata/optional_inputs_ir3.onnx:1 in f54e356. [](commit_id = f54e356b99fa00e6f3598b17c568b8188ae1df56, deletion_comment = False)

File: onnxruntime/core/graph/graph.cc [hunk 10]
```diff
@@ -2355,31 +2355,37 @@ Status Graph::SetGraphInputsOutputs() {
...
+
           if (added_input_names.end() == added_input_names.find(name)) {
             // This graph input has not been added into <graph_inputs_>.
+            bool is_initializer = name_to_initial_tensor_.find(name) != name_to_initial_tensor_.end();
+
             if (!graph_inputs_manually_set_) {
-              graph_inputs_including_initializers_.push_back(input_arg);
+              // all initializers need matching inputs if IR version < 4.
+              // for IR version >= 4 any matching inputs for initializers need to be explicitly specified
+              if (!is_initializer || ir_version_ < 4) {
```
Comment (kkaranasos) [inline onnxruntime/core/graph/graph.cc]: Maybe it's better to not hardcode this here, but instead have a (single?) util that is related to this check

File: onnxruntime/core/graph/graph_utils.cc [hunk 11]
```diff
@@ -399,9 +399,24 @@ bool IsGraphInput(const Graph& graph, const NodeArg* input) {
   return std::find(graph_inputs.begin(), graph_inputs.end(), input) != graph_inputs.end();
 }
 
+bool IsConstantInitializer(const Graph& graph, const std::string& initializer_name) {
```
Comment (kkaranasos) [inline onnxruntime/core/graph/graph_utils.cc]: I think we can/should merge IsConstantInitializer and NodeArgIsConstant, i.e., make the one call the other, since they are doing a very similar thing.

File: onnxruntime/core/framework/session_state_initializer.cc (hunk 3 above)
Comment (skottmckay) [reply onnxruntime/core/framework/session_state_initializer.cc]: I think they should be the same, but the existing handling of the TensorProto name being empty made me assume that wasn't the case. i.e. this code
    const char* name = entry.second->has_name() ? entry.second->name().c_str() : "";

As the input to this is initialized_tensor_set where the name is the key, I preferred that value. 

Now that I think about it more, I'm not sure what a TensorProto with an empty name could be used for, so I think this could be simplified to use the name from TensorProto and ignore or error out on any TensorProto where the name is empty. Does that seem reasonable?

---
In reply to: [299589384](https://github.com/microsoft/onnxruntime/pull/1324#discussion_r299589384) [](ancestors = 299589384)

File: onnxruntime/core/graph/graph_utils.cc (hunk 11 above)
Comment (skottmckay) [reply onnxruntime/core/graph/graph_utils.cc]: One is comparing pointers, one is comparing names. I left both as the pointer comparison should be faster, but maybe we don't care about such a performance difference at this point. 

---
In reply to: [299653968](https://github.com/microsoft/onnxruntime/pull/1324#discussion_r299653968) [](ancestors = 299653968)

File: onnxruntime/core/graph/graph_utils.cc (hunk 4 above)
Comment (skottmckay) [reply onnxruntime/core/graph/graph_utils.cc]: As currently implemented it's all or none. The spec says the inputs to the control flow node should match the subgraph inputs, however a model with IR version of 3 would have inputs for initializers. Due to that we first try to match all inputs (including initializers), and fallback to matching just required inputs. There's no way to partially match optional inputs as there's no information on which optional inputs are being provided.

We could possibly make it stricter if IR version is >= 4 and require exact match of all inputs.

---
In reply to: [299590945](https://github.com/microsoft/onnxruntime/pull/1324#discussion_r299590945) [](ancestors = 299590945)

File: onnxruntime/core/graph/graph_utils.cc (hunk 5 above)
Comment (skottmckay) [reply onnxruntime/core/graph/graph_utils.cc]: What can override an initializer if there's no matching graph input for it? 

---
In reply to: [299592491](https://github.com/microsoft/onnxruntime/pull/1324#discussion_r299592491) [](ancestors = 299592491)

**Comment by skottmckay** [comment]
> CodeFlow for GitHub does not download executable files. If you trust this file and would like to run it, please download it yourself through GitHub.

The general pattern is to use .onnx for a model and .pb for test input/expected output. As these are models .onnx seemed more appropriate (and provides a signal you could view it in something like netron if you wanted). 

---
In reply to: [507772523](https://github.com/microsoft/onnxruntime/pull/1324#issuecomment-507772523) [](ancestors = 507772523)

---
Refers to: onnxruntime/test/testdata/optional_inputs_ir3.onnx:1 in f54e356. [](commit_id = f54e356b99fa00e6f3598b17c568b8188ae1df56, deletion_comment = False)

File: onnxruntime/core/optimizer/constant_folding.cc (hunk 6 above)
Comment (skottmckay) [reply onnxruntime/core/optimizer/constant_folding.cc]: How is GetAllInitializedTensors changing semantics? All the initializers (overridable or not) still get added to it. 

---
In reply to: [299594145](https://github.com/microsoft/onnxruntime/pull/1324#discussion_r299594145) [](ancestors = 299594145)

File: onnxruntime/core/graph/graph.cc (hunk 10 above)
Comment (skottmckay) [reply onnxruntime/core/graph/graph.cc]: Maybe having two helpers instead would be better. 
  CanOverrideInitializer
  InitializersRequireMatchingGraphInput

Problem is, the latter one would only be used here. 

It's somewhat cryptic to have !CanOverrideInitializer() in this condition instead of ir_version_ < 4.

---
In reply to: [299648708](https://github.com/microsoft/onnxruntime/pull/1324#discussion_r299648708) [](ancestors = 299648708)

File: onnxruntime/core/framework/session_state_initializer.cc (hunk 3 above)
Comment (KeDengMS) [reply onnxruntime/core/framework/session_state_initializer.cc]: Sounds good to me.

---
In reply to: [299709883](https://github.com/microsoft/onnxruntime/pull/1324#discussion_r299709883) [](ancestors = 299709883,299589384)

File: onnxruntime/core/graph/graph_utils.cc (hunk 5 above)
Comment (KeDengMS) [reply onnxruntime/core/graph/graph_utils.cc]: I mean for this function's point of view, a name that's neither input nor initializer would be returned as true. It is better to prevent that by having some error checks.

---
In reply to: [299711239](https://github.com/microsoft/onnxruntime/pull/1324#discussion_r299711239) [](ancestors = 299711239,299592491)

File: onnxruntime/core/graph/graph_utils.cc (hunk 4 above)
Comment (KeDengMS) [reply onnxruntime/core/graph/graph_utils.cc]: If I understood correctly, seems we can add condition in graph.CanOverrideInitializer() to make subgraphs always return false.

---
In reply to: [299710986](https://github.com/microsoft/onnxruntime/pull/1324#discussion_r299710986) [](ancestors = 299710986,299590945)

File: onnxruntime/core/optimizer/constant_folding.cc (hunk 6 above)
Comment (KeDengMS) [reply onnxruntime/core/optimizer/constant_folding.cc]: OK, wrong read of the comments. Ignore me.

---
In reply to: [299713223](https://github.com/microsoft/onnxruntime/pull/1324#discussion_r299713223) [](ancestors = 299713223,299594145)

**Comment by KeDengMS** [comment]
> CodeFlow for GitHub does not download executable files. If you trust this file and would like to run it, please download it yourself through GitHub.

netron seems to treat both .pb and .onnx as onnx models. I think what you said is good to differentiate .onnx and .pb, for model and tensor separately. We may have a separate PR later to rename the .pb models in this folder.

---
In reply to: [507870064](https://github.com/microsoft/onnxruntime/pull/1324#issuecomment-507870064) [](ancestors = 507870064,507772523)

---
Refers to: onnxruntime/test/testdata/optional_inputs_ir3.onnx:1 in f54e356. [](commit_id = f54e356b99fa00e6f3598b17c568b8188ae1df56, deletion_comment = False)

**Comment by skottmckay** [comment]
> CodeFlow for GitHub does not download executable files. If you trust this file and would like to run it, please download it yourself through GitHub.

I'll go over them later today to cleanup in a separate PR

---
In reply to: [507903578](https://github.com/microsoft/onnxruntime/pull/1324#issuecomment-507903578) [](ancestors = 507903578,507870064,507772523)

---
Refers to: onnxruntime/test/testdata/optional_inputs_ir3.onnx:1 in f54e356. [](commit_id = f54e356b99fa00e6f3598b17c568b8188ae1df56, deletion_comment = False)

File: onnxruntime/core/graph/graph_utils.cc (hunk 5 above)
Comment (KeDengMS) [reply onnxruntime/core/graph/graph_utils.cc]: Glad to see the recursive search. Do we need a testcase with subgraph for it?

---
In reply to: [299744880](https://github.com/microsoft/onnxruntime/pull/1324#discussion_r299744880) [](ancestors = 299744880,299711239,299592491)

File: onnxruntime/core/graph/graph_utils.cc (hunk 4 above)
Comment (KeDengMS) [reply onnxruntime/core/graph/graph_utils.cc]: Glad to see the recursive search. Do we need a testcase with subgraph for it?

---
In reply to: [299745451](https://github.com/microsoft/onnxruntime/pull/1324#discussion_r299745451) [](ancestors = 299745451,299710986,299590945)

File: onnxruntime/core/graph/graph.cc (hunk 10 above)
Comment (skottmckay) [reply onnxruntime/core/graph/graph.cc]: As this is the only place where the check is around adding a graph input for an initializers leaving it as is.

File: onnxruntime/core/graph/graph_utils.cc (hunk 4 above)
Comment (skottmckay) [reply onnxruntime/core/graph/graph_utils.cc]: @kkaranasos Do you know if any of the optimizer or model tests will cover this?

Comment (skottmckay) [reply onnxruntime/core/graph/graph_utils.cc]: Will follow up with kkaranasos to make sure this gets covered by tests.
